react-front-end/tsrc/components/wizard/WizardHelper.tsx

```diff
@@ -0,0 +1,268 @@
+/*
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * The Apereo Foundation licenses this file to you under the Apache License,
+ * Version 2.0, (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as OEQ from "@openequella/rest-api-client";
+import * as A from "fp-ts/Array";
+import * as E from "fp-ts/Either";
+import { Eq, struct } from "fp-ts/Eq";
+import { absurd, flow, pipe } from "fp-ts/function";
+import * as N from "fp-ts/number";
+import * as O from "fp-ts/Option";
+import { Refinement } from "fp-ts/Refinement";
+import * as S from "fp-ts/string";
+import * as React from "react";
+import { WizardEditBox } from "./WizardEditBox";
+import { WizardUnsupported } from "./WizardUnsupported";
+
+/**
+ * Used to loosely target what a value (typically a `ControlValue`) is being used for.
+ */
+export interface ControlTarget {
+  /**
+   * The 'fullPath's for the targetNode.
+   */
+  schemaNode: string[];
+  /**
+   * The type of control that is being targeted.
+   */
+  type: OEQ.WizardControl.ControlType;
+}
+
+/**
+ * Convenience type for our way of storing the two main value types across our controls. Represents
+ * that some controls are textual, and some are numeric; and that some controls store more than one
+ * value.
+ */
+export type ControlValue = number[] | string[];
+
+/**
+ * Identifies a Wizard 'field' and specifies its value.
+ */
+export interface FieldValue {
+  target: ControlTarget;
+  value: ControlValue;
+}
+
+/**
+ * Check the type of the head of an array. Returns `false` if empty.
+ *
+ * @param refinement method to validate head element.
+ */
+const isHeadType =
+  <T,>(refinement: Refinement<unknown, T>) =>
+  (xs: unknown[]): boolean =>
+    pipe(
+      xs,
+      A.head,
+      O.map(refinement),
+      O.getOrElse((): boolean => false)
```

Comment:
Ah this would've been more concise (save function definition) and would've had given the typing grief if I had've used the fp-ts `function` util of `constFalse`. Something like:

```suggestion
      O.getOrElse(constFalse)
```